change internal extend to 16384 for better rendering quality

candux opened the pull request:
# Terminology
- MVT-Extent: the width and height of a MVT-Tile in integer coordinates. See the [spec](https://github.com/mapbox/vector-tile-spec/tree/master/2.1#41-layers) for more details
- Maplibre-Extent: the width and height of the internal coordinate system for each tile inside maplibre.


# Background
Most MVT Datasets are limited to zoom-level 14. Generating higher zoom levels for large areas is very demanding. If you want to keep more details in MVT while staying on the same zoom-level, it is possible to increase the MVT-extent of layers for better overzooming. This became very easy to do with _ST_AsMVT_. Since Maplibre uses an internal extent of 8192, details of bigger MVT-Extents get lost.

# PR
In this PR the internal extent of Maplibre is doubled from 8192 to 16384. This improves the quality of layers with a MVT-Extent greater 8192. This is done by using one more bit of the 16-Bit value, passed to WebGL. This bit was used so far only as an hack in the line layer, to pass additional attributes. This hack is removed in the first commit.

# Screenshots
This shows the difference of main and this PR on Tiles with max-zoom 14, MVT-Extent of 16384 on Zoom-Level 21.
Before: 
![libre_orig](https://user-images.githubusercontent.com/7271757/169890786-1b10b669-c9ac-4b01-9fcf-c6c99fbd5138.png)
After:
![libre_optim](https://user-images.githubusercontent.com/7271757/169890817-64f1cfc7-24c4-4753-bac0-ecef081168d6.png)
The building is much more detailed and lines are more parallel.

# Performance
Increasing the extent itself should not mess with the performance in any way. 
Removing the hack in the line layer can have an impact on performance. There are two additional attributes passed to WebGL. On the other hand, the shader has a few instructions less. I wasn't able to measure any real difference because the line performance test is too noisy.
![performance_extent](https://user-images.githubusercontent.com/7271757/169891481-db5af916-961b-47b5-b2b7-06c845436624.png)

# Status
This PR is at the moment not quite ready to be merged. I still have to run all tests and make the same changes in maplibre-gl-native to stay compatible with the shaders. I would like to ask for feedback, if this change is welcome and if there are more things I have to do.

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->


 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Post benchmark scores.
 - [x] Manually test the debug page.
 - [x] Suggest a changelog category: bug/feature/docs/etc. or "skip changelog".: feature
 - [x] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.
 - [x] Update rendering tests
 - [x] make the same change in maplibre-gl-native: https://github.com/maplibre/maplibre-gl-native/pull/289

Commit:
don't use bit in coordinates for up and round attributes

this effectively reverts https://github.com/mapbox/mapbox-gl-js/pull/8306

Commit:
use 14 bits for internal coordinates

HarelM commented:
There were a few changes recently to the sharers including terrain 3D which are missing in the native code, so I'm not sure you be able to test the sharers.
I personally don't see the difference in the pictures, but the code looks a lot cleaner. :-)

Commit:
update unit tests for new results

wipfli commented:
Just out of curiosity, what are your use-cases @candux for zoom levels higher than 14?

nyurik commented:
> Just out of curiosity, what are your use-cases @candux for zoom levels higher than 14?

@wipfli this doesn't change the typical z14 for datasets / z18+ for visualizations. This PR only makes overzooming (z15+) crispier.

Commit:
fix fill_extrusion_programm

this fixes all wrong patterns. Unfortunatly I really don't know where
the magic number comes from

Commit:
Merge remote-tracking branch 'old-origin/main' into david/quality

candux commented:
> There were a few changes recently to the sharers including terrain 3D which are missing in the native code, so I'm not sure you be able to test the sharers.

Thanks for the heads-up. I try to prepare everything in https://github.com/maplibre/maplibre-gl-native/pull/289 , so it can just be merged when the shaders in native get updated to the latest version of js.

> I personally don't see the difference in the pictures, but the code looks a lot cleaner. :-)

The changes are most visible in the building. The lines are crispier and more parallel.

> Just out of curiosity, what are your use-cases @candux for zoom levels higher than 14?

We are mostly interested to show more indoor details of train-stations

> this doesn't change the typical z14 for datasets / z18+ for visualizations. This PR only makes overzooming (z15+) crispier.

Yes, exactly. We want to avoid creating a new dataset just for indoor visualization which complicates the style and exporting the whole world on z15+ is quite difficult.
  
I currently work on fixing all rendering regressions. Unfortunately there are a few places which use magic numbers which depend on the extent. Refactoring these to use the globally defined constant would be a great maintenance improvement for the future.

Commit:
improve geoJson source quality

makes now sense with the greater extent

Commit:
update query results for better precision

Commit:
update render results

as far as I can tell all differences are because of better precision

wipfli commented:
I think it would be a good moment to put the magic number 8192 or its new value into a unchangeable global variable or something like that.

prozessor13 commented:
One sideeffect of the greater extent would be, that the vectortiles will increase in size, because in protobuf-format bigger numbers will use more bytes. I dont know if this effect is major or minor, do you? In general i like this PR!

prozessor13 commented:
No, only on the protobuf vectortiles.

candux commented:
> I think it would be a good moment to put the magic number 8192 or its new value into a unchangeable global variable or something like that.

it's already in https://github.com/maplibre/maplibre-gl-js/pull/1244/files#diff-27fcda492afcb31d627de1fdef3e3dd88eac4e122ed6cc2fb8fe7b4b2e0bee9f

> One sideeffect of the greater extent would be, that the vectortiles will increase in size, because in protobuf-format bigger numbers will use more bytes. I dont know if this effect is major or minor, do you? In general i like this PR!

we increased the tile extent from 4094 to 16384 for some openmaptiles layers (buildings, transportation). The size increase was only a few percents, but the quality improved a lot

Commit:
remove outdated comment

Commit:
fix comment

Commit:
fix flipping of vertical lines

Commit:
add changelog entry

Commit:
update results of text-pitch-alignment/viewport-text-rotation-alignment-map

only minor differences because of preciser coordinates

Commit:
update results of viewport-text-depthtest render test

the text below the red line is what is expected from the style.
It already looks like that on main, but the diff wasn't big enough to
trigger a fail.
Minor changes of the position pushed to the test to fail

candux commented:
Finally, all tests pass. A lot of results had to be updated but it should always be justified because of better precision.
viewport-text-depthtest looks like a change in behavior but the new result is correct.

From my perspective, this is ready to be merged.
I would be very happy if someone could review or test this PR.

wipfli commented:
Can you post the benchmark results? I recently fixed them so maybe you need to merge main first...

wipfli commented:
Please comment on the apparent difference in test/integration/render/tests/text-pitch-alignment/viewport-text-depthtest/expected.png

wipfli commented:
@candux have you use one of the native platforms in the past? I would like to be able to update native when the shaders in GL JS change. See https://github.com/maplibre/maplibre-gl-native/pull/315 

Any help on the native stuff would be really helpful for pull requests that change the shaders in GL JS.

candux commented:
> Can you post the benchmark results? I recently fixed them so maybe you need to merge main first...

Unfortunatly they always get stuck for me at some point. This also happens on main. Doesn't matter if they run in the browser or headless. I will try it on another system.

> Please comment on the apparent difference in test/integration/render/tests/text-pitch-alignment/viewport-text-depthtest/expected.png

I put the explanation in the git commit:
> the text below the red line is what is expected from the style.
>It already looks like that on main, but the diff wasn't big enough to
>trigger a fail.
> Minor changes of the position pushed to the test to fail

I can also make new PR with just that commit to update the expected results

> have you use one of the native platforms in the past? I would like to be able to update native when the shaders in GL JS change. See https://github.com/maplibre/maplibre-gl-native/pull/315
> Any help on the native stuff would be really helpful for pull requests that change the shaders in GL JS.

Unfortuantly, I'm not familiar with native and only use it in tileserver-gl. I will take a look if I can help but that not really my expertise.

wipfli commented:
Would be great if you could update the render test in a separate pull request. Thanks for catching this.

wipfli commented:
I can give the benchmarks also a try and post them here...

wipfli committed:
Merge branch 'main' into david/quality

wipfli commented:
On a first benchmark run on my laptop, http://0.0.0.0:9966/bench/versions?compare=main#WorkerTransfer seems roughly 10 percent slower. @HarelM is that the one which is always 10 percent slower in the second run?

wipfli commented:
I merged main, now the benchmarks work again...

HarelM commented:
In order to avoid issues with benchmarks I run 3 version comparison, the first version is "dummy" in order to make the comparison good between the second and the third version.

HarelM commented:
Where are we with this PR? What's keeping us from merging this?

wipfli commented:
I wonder what the implications for native are when we change the shaders...

HarelM commented:
Yes, I agree. We should discuss this in our next steering committee meeting, as the GL-JS is far ahead of the native when it comes to the shaders and we need to decide what to do there...

wipfli commented:
> This bit was used so far only as an hack in the line layer, to pass additional attributes. This hack is removed in the first commit.

Maybe this bit hack was a performance optimization. @candux what do you think, is it possible that with this pull request the rendering gets slower?

birkskyum commented:
Stale or not - don't auto-close. We need to get maplibre-gl-js is updated in maplibre-gl-native, and the benchmarks fixed, in order to get a better picture of the performance implications of this, because this visual improvement would be great to have.

HarelM commented:
It's worth mentioning that native is no longer using the shaders that are here.
So there is no block from this part.
cc @ChrisLoer

birkskyum commented:
@candux interesting - if the other solution is in development, I think it's best to land it first and reassess the situation. I got reminded of this ticket again yesterday when I tried to add a point to an intersection - and it seemed like the level of accuracy needed to get it right simply wasn't supported.

<img width="220" alt="Screenshot 2023-06-01 at 04 40 41" src="https://github.com/maplibre/maplibre-gl-js/assets/74932975/b129ffd2-ecec-417c-a048-308bdae74e26">

ChrisLoer commented:
> if the other solution is in development, I think it's best to land it first and reassess the situation.

It's unfortunately not in active development -- I made a prototype and want to do it, but don't have available time to work on it right now. Felt needs a solution, so we're motivated to find someone to work on it, but it's still possible we'll do a server-side solution as the expedient path.

birkskyum commented:
@ChrisLoer , would you consider this PR an improvement to the situation until the other PR eventually lands?

ChrisLoer commented:
> @ChrisLoer , would you consider this PR an improvement to the situation until the other PR eventually lands?

Yeah, it's an improvement -- but wouldn't fundamentally change our need for a fuller overzooming solution. If we assume "more than 6 levels of overzooming starts to look bad", then with this change it becomes "more than 7 levels of overzooming looks bad". It also doesn't help with the line-labeling issue (too many line-labels generated on overzoomed tiles, slowing things down and eventually leading to glitches or even crashes).

I don't think there's any incompatibility between the two approaches though, right? One thing that occurs to me is that if we implemented https://github.com/maplibre/maplibre-gl-js/issues/2507 with this also in place, we could get away with waiting an extra zoom level before triggering the on-the-fly tile generation, which would presumably be a perf win.

candux commented:
@ChrisLoer We (https://geops.com/en) would be able to work on https://github.com/maplibre/maplibre-gl-js/issues/2507 and/or finish this PR.  Do you think Felt could be interested to finance this work?
